test/utils/InstanceTest.sol

```diff
@@ -0,0 +1,451 @@
+    ///      deployer coordinator contract.
+    /// @param deploymentId The deployment id.
+    /// @param deploymentSalt The deployment salt for create2.
+    /// @param contribution The amount to initialize the market.
+    /// @param asBase Initialize the market with base token.
+    function deployHyperdrive(
+        bytes32 deploymentId,
+        bytes32 deploymentSalt,
+        uint256 contribution,
+        bool asBase
+    ) private {
+        // Alice is the default deployer.
+        vm.startPrank(alice);
+
+        // Deploy Hyperdrive target contracts.
+        factory.deployTarget(
+            deploymentId,
+            deployerCoordinator,
+            poolConfig,
+            new bytes(0),
+            FIXED_RATE,
+            FIXED_RATE,
+            0,
+            deploymentSalt
+        );
+        factory.deployTarget(
+            deploymentId,
+            deployerCoordinator,
+            poolConfig,
+            new bytes(0),
+            FIXED_RATE,
+            FIXED_RATE,
+            1,
+            deploymentSalt
+        );
+        factory.deployTarget(
+            deploymentId,
+            deployerCoordinator,
+            poolConfig,
+            new bytes(0),
+            FIXED_RATE,
+            FIXED_RATE,
+            2,
+            deploymentSalt
+        );
+        factory.deployTarget(
+            deploymentId,
+            deployerCoordinator,
+            poolConfig,
+            new bytes(0),
+            FIXED_RATE,
+            FIXED_RATE,
+            3,
+            deploymentSalt
+        );
+        factory.deployTarget(
+            deploymentId,
+            deployerCoordinator,
+            poolConfig,
+            new bytes(0),
+            FIXED_RATE,
+            FIXED_RATE,
+            4,
+            deploymentSalt
+        );
+
+        // Alice gives approval to the deployer coordinator to fund the market.
+        config.token.approve(deployerCoordinator, 100_000e18);
+
+        // We expect the deployAndInitialize to fail with an
+        // Unsupported token error if depositing with base is not supported.
+        // If the base token is ETH we expect a NotPayable error.
+        if (!config.enableBaseDeposits && asBase) {
+            vm.expectRevert(
+                isBaseETH
+                    ? IHyperdrive.NotPayable.selector
+                    : IHyperdrive.UnsupportedToken.selector
+            );
+        }
+
+        // We expect the deployAndInitialize to fail with an
+        // Unsupported token error if depositing with shares is not supported.
+        if (!config.enableShareDeposits && !asBase) {
+            vm.expectRevert(IHyperdrive.UnsupportedToken.selector);
+        }
+
+        // Deploy and initialize the market. If the base token is ETH we pass the
+        // contribution through the call.
+        hyperdrive = factory.deployAndInitialize{
+            value: asBase && isBaseETH ? contribution : 0
```

Comment:
It looks like we aren't testing refunds here. I think it would be good to add a check for this since each deployer coordinator has to implement their own refunds. Handling ETH can be a bit of a footgun.